### PR TITLE
Add get config functions

### PR DESCRIPTION
**Pull request by dyates**
Adds the ability to retrieve a `QuantumProcessorConfig` from a project.  There are two ways to retrieve processor configs: from the `Engine` or an `EngineProcessor`.  

Here is an example using the processor:

```
engine = cirq_google.get_engine('i-dont-have-experimental')
processor = engine.get_processor('WIL01_9A_BLACK')

run_name_config = processor.get_config_by_run_name(run_name="abc", config_id="123")
current_config = processor.get_config_by_run_name(config_id="123")
snapshot_config = processor.get_config_by_snapshot_id(snapshot_id="xyz", config_id="123")
```

Here is an example getting the config directly from the Engine:

```
engine = cirq_google.get_engine('i-dont-have-experimental')
processor_id = 'WIL01_9A_BLACK'

run_name_config = engine.get_config_by_run_name(
    processor_id=processor_id, run_name="abc", config_id="123"
)
current_config = engine.get_config_by_run_name(
    processor_id=processor_id, config_id="123"
)
snapshot_config = engine.get_config_by_snapshot_id(
    processor_id=processor_id, snapshot_id="xyz", config_id="123"
)
```

### CODE REVIEW

**Comment by dyates** [comment]
I'm working on fixing the errors.  Some missing tests/logic for a few edge cases.

File: cirq-google/cirq_google/cloud/quantum/__init__.py [hunk 1]
```diff
@@ -13,6 +13,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
+from cirq_google.cloud.quantum import gapic_version as package_version
+
+__version__ = package_version.__version__
+
```
Comment (hoisinberg) [inline cirq-google/cirq_google/cloud/quantum/__init__.py]: These imports and the `gapic_version.py` files they reference aren't actually needed and should be removed.

File: cirq-google/cirq_google/engine/abstract_engine.py [hunk 2]
```diff
@@ -134,3 +139,35 @@ def get_sampler(self, processor_id: str | list[str]) -> cirq.Sampler:
...
+    @abc.abstractmethod
+    def get_processor_config_by_snapshot_id(
+        self, processor_id: str, snapshot_id: str, config_id: str
+    ) -> abstract_processor_config.AbstractProcessorConfig | None:
+        """Returns a ProcessorConfig from this project and the given processor id.
+
+        Args:
+           processor_id: The processor unique identifier.
+           snapshot_id: The unique identifier for the snapshot.
+           config_id: The unique identifier for the snapshot.
```
Comment (hoisinberg) [inline cirq-google/cirq_google/engine/abstract_engine.py]: "snapshot" -> "config"

File: cirq-google/cirq_google/engine/abstract_engine.py [hunk 3]
```diff
@@ -134,3 +139,35 @@ def get_sampler(self, processor_id: str | list[str]) -> cirq.Sampler:
...
+        self, processor_id: str, config_id: str, run_name: str = 'current'
+    ) -> abstract_processor_config.AbstractProcessorConfig | None:
+        """Returns a ProcessorConfig from this project and the given processor id.
+
+        If no run_name is provided, the config from the most recent run is returned.
+
+        Args:
+            processor_id: The processor unique identifier.
+            run_name: The unique identifier for the automation run.
+            config_id: The unique identifier for the snapshot.
```
Comment (hoisinberg) [inline cirq-google/cirq_google/engine/abstract_engine.py]: "snapshot" -> "config"

File: cirq-google/cirq_google/engine/processor_config.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    Raise:
+        ValueError: If quantum_processor_config is contains incompatible types.
+    """
+
+    def __init__(
+        self, *, quantum_processor_config: quantum.QuantumProcessorConfig, run_name: str = ''
+    ) -> None:
+        self._quantum_processor_config = quantum_processor_config
+        self._run_name = run_name
+        self._device_spec = util.unpack_any(
```
Comment (hoisinberg) [inline cirq-google/cirq_google/engine/processor_config.py]: You can save the GridDevice constructed from this proto as a member instead, to remove the need to reconstruct it on each call to `effective_device`. Likewise with `calibration`.

File: cirq-google/cirq_google/engine/abstract_engine.py [hunk 5]
```diff
@@ -134,3 +139,35 @@ def get_sampler(self, processor_id: str | list[str]) -> cirq.Sampler:
...
+           snapshot_id: The unique identifier for the snapshot.
+           config_id: The unique identifier for the snapshot.
+
+        Returns:
+           The ProcessorConfig from this project and processor.
+        """
+
+    @abc.abstractmethod
+    def get_processor_config_by_run_name(
+        self, processor_id: str, config_id: str, run_name: str = 'current'
```
Comment (hoisinberg) [inline cirq-google/cirq_google/engine/abstract_engine.py]: After some recent discussions about how to handle default run_names, we've actually settled on a default `run_name` of `"default"`. To be replaced here and elsewhere.

`config_alias` should have a default value of `"default"` and should be placed after `run_name` to maintain consistency with `get_config_by_snapshot_id`.

File: cirq-google/cirq_google/engine/abstract_engine.py [hunk 6]
```diff
@@ -134,3 +139,35 @@ def get_sampler(self, processor_id: str | list[str]) -> cirq.Sampler:
             processor_id: String identifier, or list of string identifiers,
                 determining which processors may be used when sampling.
         """
+
+    @abc.abstractmethod
+    def get_processor_config_by_snapshot_id(
+        self, processor_id: str, snapshot_id: str, config_id: str
```
Comment (hoisinberg) [inline cirq-google/cirq_google/engine/abstract_engine.py]: In order to maintain consistency with how this is called in other parts of the stack, let's rename `config_id` to `config_alias` here and elsewhere. 

`config_alias` should also have a default value of `"default"`. Because this means "default" doesn't uniquely identify a config, the documentation should just say something like "Identifier for the config to retrieve".

File: cirq-google/cirq_google/engine/abstract_engine.py [hunk 7]
```diff
@@ -134,3 +139,35 @@ def get_sampler(self, processor_id: str | list[str]) -> cirq.Sampler:
             processor_id: String identifier, or list of string identifiers,
                 determining which processors may be used when sampling.
         """
+
+    @abc.abstractmethod
+    def get_processor_config_by_snapshot_id(
```
Comment (hoisinberg) [inline cirq-google/cirq_google/engine/abstract_engine.py]: The naming of these methods seems to be a bit inconsistent across the different files. Let's standardize to `get_processor_config_from_snapshot` and `get_processor_config_from_run`. The "processor" bit can be omitted from the `AbstractProcessor` and its subclasses.

File: cirq-google/cirq_google/engine/abstract_processor_config.py
Comment (hoisinberg) [inline cirq-google/cirq_google/engine/abstract_processor_config.py:1]: I don't think we really need an abstract base class for this, since a `ProcessorConfig` is really only meant to be a data object.

File: dev_tools/notebooks/isolated_notebook_test.py [hunk 8]
```diff
@@ -69,6 +69,8 @@
     # temporary: need to fix QVM metrics and device spec
     'docs/tutorials/google/spin_echoes.ipynb',
     'docs/tutorials/google/visualizing_calibration_metrics.ipynb',
+    # temporary: allow name changes of keyword arguments in quantum engine interfaces to stabilize
+    'docs/simulate/virtual_engine_interface.ipynb',
```
Comment (hoisinberg) [inline dev_tools/notebooks/isolated_notebook_test.py]: This has since been removed. Looks like you may just need to merge from main to get rid of this.

**Review by hoisinberg — CHANGES_REQUESTED** [review]

File: cirq-google/cirq_google/engine/abstract_engine.py (hunk 5 above)
Comment (dyates) [reply cirq-google/cirq_google/engine/abstract_engine.py]: Done

File: cirq-google/cirq_google/engine/processor_config.py (hunk 4 above)
Comment (dyates) [reply cirq-google/cirq_google/engine/processor_config.py]: Done.

File: cirq-google/cirq_google/engine/abstract_engine.py (hunk 3 above)
Comment (dyates) [reply cirq-google/cirq_google/engine/abstract_engine.py]: Done

File: cirq-google/cirq_google/engine/abstract_engine.py (hunk 2 above)
Comment (dyates) [reply cirq-google/cirq_google/engine/abstract_engine.py]: Done

File: cirq-google/cirq_google/cloud/quantum/__init__.py (hunk 1 above)
Comment (dyates) [reply cirq-google/cirq_google/cloud/quantum/__init__.py]: Reomved

File: cirq-google/cirq_google/cloud/quantum/gapic_version.py
Comment (hoisinberg) [inline cirq-google/cirq_google/cloud/quantum/gapic_version.py:1]: Remove

File: cirq-google/cirq_google/cloud/quantum_v1alpha1/__init__.py [hunk 9]
```diff
@@ -13,6 +13,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
+from . import gapic_version as package_version
+
+__version__ = package_version.__version__
+
```
Comment (hoisinberg) [inline cirq-google/cirq_google/cloud/quantum_v1alpha1/__init__.py]: Remove

File: cirq-google/cirq_google/cloud/quantum_v1alpha1/gapic_version.py
Comment (hoisinberg) [inline cirq-google/cirq_google/cloud/quantum_v1alpha1/gapic_version.py:1]: Remove

File: cirq-google/cirq_google/engine/abstract_engine.py [hunk 10]
```diff
@@ -134,3 +139,35 @@ def get_sampler(self, processor_id: str | list[str]) -> cirq.Sampler:
...
+           The ProcessorConfig from this project and processor.
+        """
+
+    @abc.abstractmethod
+    def get_processor_config_from_run(
+        self, processor_id: str, run_name: str = 'default', config_alias: str = 'default'
+    ) -> processor_config.ProcessorConfig | None:
+        """Returns a ProcessorConfig from this project and the given processor id.
+
+        If no run_name is provided, the config from the most recent run is returned.
```
Comment (hoisinberg) [inline cirq-google/cirq_google/engine/abstract_engine.py]: I think you can remove this line now that we have the default argument for `run_name`, or mention something about how we will use an internally-configured default run if left unspecified. Here and in the overrides.

File: cirq-google/cirq_google/engine/engine.py [hunk 11]
```diff
@@ -635,6 +636,61 @@ def get_sampler(
...
+        return None
+
+    get_processor_config_from_snapshot = duet.sync(get_processor_config_from_snapshot_async)
+
+    async def get_processor_config_from_run_async(
+        self, processor_id: str, run_name: str = 'current', config_alias: str = 'default'
+    ) -> processor_config.ProcessorConfig | None:
+        """Returns a ProcessorConfig from this project and the given processor id.
+
+        If no run_name is provided, the config from the most recent run is returned.
```
Comment (hoisinberg) [inline cirq-google/cirq_google/engine/engine.py]: Update to talk about the default args

File: cirq-google/cirq_google/engine/engine_processor.py [hunk 12]
```diff
@@ -497,6 +503,60 @@ def get_schedule(
         filter_str = ' AND '.join(filters)
         return self.context.client.list_time_slots(self.project_id, self.processor_id, filter_str)
 
+    def get_config_from_run(
+        self, run_name: str = 'default', config_alias: str = 'default'
+    ) -> processor_config.ProcessorConfig | None:
+        """Retrieves a ProcessorConfig from an automation run.
+
+            If no run name is provided, the config from the most recent run
```
Comment (hoisinberg) [inline cirq-google/cirq_google/engine/engine_processor.py]: Update to mention default args

File: cirq-google/cirq_google/engine/processor_config.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+    @property
+    def calibration(self) -> cg.Calibration:
+        """Charicterization metrics captured for this configuration"""
+        return self._calibration
+
+    @property
+    def snapshot_id(self) -> str:
+        """The snapshot that contains this processor config"""
+        if 'configSnapshots' not in self._quantum_processor_config.name:
+            # We assume the calling `get_quantume_processor_config` always
```
Comment (hoisinberg) [inline cirq-google/cirq_google/engine/processor_config.py:71]: quantume -> quantum

File: cirq-google/cirq_google/engine/abstract_engine.py [hunk 14]
```diff
@@ -134,3 +139,35 @@ def get_sampler(self, processor_id: str | list[str]) -> cirq.Sampler:
             processor_id: String identifier, or list of string identifiers,
                 determining which processors may be used when sampling.
         """
+
+    @abc.abstractmethod
+    def get_processor_config_from_snapshot(
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/abstract_engine.py]: I don't think we want to enforce implementation of these methods across all `AbstractEngine` implementations - most of these are simulators and "quantum virtual machines" and don't have a notion of a variable configuration.

File: cirq-google/cirq_google/engine/abstract_processor.py [hunk 15]
```diff
@@ -377,3 +378,35 @@ def get_schedule(
         Returns:
             Schedule time slots.
         """
+
+    @abc.abstractmethod
+    def get_config_from_run(
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/abstract_processor.py]: As for the abstract_engine, these probably aren't needed for the abstract processors and can just be moved to `EngineProcessor`.

File: cirq-google/cirq_google/engine/processor_config.py [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+from cirq_google.api import v2
+from cirq_google.engine import util
+
+if TYPE_CHECKING:
+    import cirq
+    from cirq_google.cloud.quantum_v1alpha1.types import quantum
+
+
+class ProcessorConfig:
+    """Representation of a quantum processor configuration
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/processor_config.py]: Nit: add punctuation.

File: cirq-google/cirq_google/engine/processor_config.py [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+
+    Describes available qubits, gates, and calivration data associated with
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/processor_config.py]: s/calivration/calibration

File: cirq-google/cirq_google/engine/processor_config.py [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+    a processor configuration.
+
+    Raise:
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/processor_config.py]: This docstring is for the class, but the `Raise` clause here looks like it should be documented on the constructor.

File: cirq-google/cirq_google/engine/processor_config.py [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+            # We assume the calling `get_quantume_processor_config` always
+            # returns a config with the snapshot resouce nanme.  This check
+            # is added in case this behavior changes in the future.
+            return ''
+        parts = self._quantum_processor_config.name.split('/')
+        return parts[5]
+
+    @property
+    def run_name(self) -> str:
+        """The run that generated this config if avaiable."""
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/processor_config.py:80]: This should also document the value returned when it's not available.

File: cirq-google/cirq_google/engine/processor_config.py [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+        )
+        self._calibration = cg.Calibration(
+            util.unpack_any(
+                self._quantum_processor_config.characterization, v2.metrics_pb2.MetricsSnapshot()
+            )
+        )
+
+    @property
+    def effective_device(self) -> cirq.Device:
+        """The GridDevice generated from thisc configuration's device specification"""
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/processor_config.py:59]: s/thisc/this

Also - add punctuation (here and below).

File: cirq-google/cirq_google/engine/processor_config.py [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+        return self._grid_device
+
+    @property
+    def calibration(self) -> cg.Calibration:
+        """Charicterization metrics captured for this configuration"""
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/processor_config.py:64]: s/Charicterization/Characterization

File: cirq-google/cirq_google/engine/processor_config.py [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+        return self._calibration
+
+    @property
+    def snapshot_id(self) -> str:
+        """The snapshot that contains this processor config"""
+        if 'configSnapshots' not in self._quantum_processor_config.name:
+            # We assume the calling `get_quantume_processor_config` always
+            # returns a config with the snapshot resouce nanme.  This check
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/processor_config.py:72]: s/resouce nanme/resource name

File: cirq-google/cirq_google/engine/processor_config.py [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+
+    @property
+    def config_alias(self) -> str:
+        """The unique identifier for this config."""
+        parts = self._quantum_processor_config.name.split('/')
+        return parts[-1]
+
+    def __repr__(self) -> str:
+        return (
+            f'cirq_google.ProcessorConfig'
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/processor_config.py]: I don't think fstring is needed here since there's no interpolation.

File: cirq-google/cirq_google/engine/processor_config.py [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+        return parts[1]
+
+    @property
+    def processor_id(self) -> str:
+        """The processor id for this config."""
+        parts = self._quantum_processor_config.name.split('/')
+        return parts[3]
+
+    @property
+    def config_alias(self) -> str:
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/processor_config.py]: We call this "config_name" the Quantum Engine API. @hoisinberg - what do you think about converging on the "config_name" over the older "config_alias" as we update all of this config tooling?

File: cirq-google/cirq_google/engine/processor_config.py [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+        parts = self._quantum_processor_config.name.split('/')
+        return parts[5]
+
+    @property
+    def run_name(self) -> str:
+        """The run that generated this config if avaiable."""
+        return self._run_name
+
+    @property
+    def project_id(self) -> str:
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/processor_config.py]: Do we expect anyone to use this? It strikes me as more of an implementation detail and something that we probably don't need to expose.

File: cirq-google/cirq_google/engine/engine_processor.py [hunk 26]
```diff
@@ -497,6 +503,60 @@ def get_schedule(
...
+
+        Args:
+            processor_id: The processor unique identifier.
+            config_alias: The quantum processor's unique identifier.
+            run_name: The automation run name.  Use 'default'
+                      if none id provided.
+
+        Returns: The quantum processor config.
+        """
+        response = self.context.client.get_quantum_processor_config_from_run(
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/engine_processor.py]: WDYT of forwarding to the Engine implementation to ensure that both versions implement the same functionality, e.g. what to call, how to transform the response? e.g., that's also done for `run_sweep`.

**Comment by hoisinberg** [comment]
I've verified that this works fully end-to-end with a colab containing the following code:

```
import cirq
import cirq_google
from cirq_google.engine.qcs_notebook import authenticate_user

authenticate_user()
engine = cirq_google.get_engine('quantum-michaelbq')
processor_id = 'WIL01_9A_BLACK'
run_name = 'current'
config_alias = 'cz_8'

print("Get config from engine")
config = engine.get_processor_config_from_run(processor_id=processor_id, run_name=run_name, config_alias=config_alias)
print(config.effective_device)

print("Get config from processor object")
config = engine.get_processor(processor_id).get_config_from_run(run_name=run_name, config_alias=config_alias)
print(config.effective_device)
```

File: cirq-google/cirq_google/engine/processor_config.py (hunk 24 above)
Comment (hoisinberg) [reply cirq-google/cirq_google/engine/processor_config.py]: Standardizing to config_name sounds good to me.

File: cirq-google/cirq_google/engine/abstract_local_engine.py [hunk 27]
```diff
@@ -175,3 +175,15 @@ def get_sampler(self, processor_id: str | list[str]) -> cirq.Sampler:
         if not isinstance(processor_id, str):
             raise ValueError(f'Invalid processor {processor_id}')
         return self._processors[processor_id].get_sampler()
+
+    def get_processor_config_from_snapshot(
+        self, processor_id: str, snapshot_id: str, config_alias: str = 'default'
```
Comment (hoisinberg) [inline cirq-google/cirq_google/engine/abstract_local_engine.py]: Final nit: let's change `config_alias` to `config_name` here and throughout.

**Review by hoisinberg — APPROVED** [review]
Looks good on my part. Looks like there are just a couple of comments left from Will.

File: cirq-google/cirq_google/engine/abstract_local_engine_test.py [hunk 28]
```diff
@@ -81,6 +81,12 @@ def list_programs(
     def get_program(self, program_id: str) -> AbstractProgram:
         return self._programs[program_id]
 
+    def get_config_from_run(self, *args, **kwargs):
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/abstract_local_engine_test.py]: Is this needed? It's no longer part of the `AbstractLocalProcessor` afaict.

File: cirq-google/cirq_google/engine/engine_client.py [hunk 29]
```diff
@@ -1230,6 +1300,28 @@ def _ids_from_calibration_name(calibration_name: str) -> tuple[str, str, int]:
     return parts[1], parts[3], int(parts[5])
 
 
+def _quantum_processor_name_with_snapshot_id(
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/engine_client.py]: This name represents a "processor config" and not a "processor"

Comment (wcourtney) [inline cirq-google/cirq_google/engine/engine_client.py]: There is an existing pattern above of "<resource>_name_from_ids" - can we use that instead for consistency?

File: cirq-google/cirq_google/engine/engine_client.py [hunk 30]
```diff
@@ -1230,6 +1300,28 @@ def _ids_from_calibration_name(calibration_name: str) -> tuple[str, str, int]:
...
+    project_id: str, processor_id: str, snapshot_id: str, config_name: str
+) -> str:
+    return (
+        f'projects/{project_id}/'
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/engine_client.py]: Nit: since this is a sub-resource of processor, WDYT of relying on `_processor_name_from_ids()` to generate the processor name from `project` and `processor_id` and then only appending the config context here?

**Review by wcourtney — CHANGES_REQUESTED** [review]
Overall LGTM % a few remaining nits.

File: cirq-google/cirq_google/engine/abstract_local_engine.py (hunk 27 above)
Comment (dyates) [reply cirq-google/cirq_google/engine/abstract_local_engine.py]: Removed.

**Review by wcourtney — APPROVED** [review]
Thanks for the updates!
